### ISSUE TITLE
POST index/_settings with body produces wrong exception

### ISSUE THREAD

**Issue by HonzaKral** [elastic/elasticsearch#17853]
**Elasticsearch version**: 2.1 - master

**Description of the problem including expected versus actual behavior**: Sending even an empty body (`url localhost:9200/test/_settings -d ''` or curl localhost:9200/test/_settings -d '{}') results in a (wrong) exception - `mapper_parsing_exception`

**Steps to reproduce**:
1. `curl -X PUT localhost:9200/test`
2. `curl localhost:9200/test/_settings -d ''`

**Comment by nikoncode** [comment]
What is expected behavior?

**Comment by HonzaKral** [comment]
Either for it to ignore the body and respond with the settings or provide a relevant exception saying that sending body is not allowed with this API

**Comment by clintongormley** [comment]
It's a POST request, not a GET request, and it is somehow matching a `POST /{index}/_mapping` handler

**Comment by alexshadow007** [comment]
@clintongormley 
It is matching `POST /{index}/{type}` handler.
It is trying to create dynamic mapping for type `_settings`, but can't and throws MapperParsingException.

**Comment by jeff303** [comment]
I'm new to the project, so take everything here with a huge grain of salt.  It's not clear to me how responding to the POST the same way as to the GET, or changing the exception could be accomplished cleanly.

`GET /{index}/_settings` is mapped in `RestGetIndicesAction`, with the _settings portion being covered by a feature.  In theory, a specific POST handler could be added for `/{index}/_settings` but it probably shouldn't go in this class, since it's explicitly for GET requests.

Absent a special handler for this, even changing the exception would require some kind of special handling in `RestIndexAction`.  But it would seem that prevents someone from dynamically indexing a document with type `_settings`, and why shouldn't that be allowed?

**Comment by SushantVarshney** [comment]
Also facing the same issue.Is there any way to overcome this?

**Comment by qwerty4030** [comment]
@clintongormley @jbertouch I think this may be user error. curl and sense automatically convert a `GET` request with a body to a `POST`. I was able to reproduce the error on latest master, but changing the request to the following gives the correct index settings (request body ignored):
`curl -X GET localhost:9200/test/_settings -d ''`

**Comment by jasontedor** [comment]
> I think this may be user error.

Whether or not it's user error, there is the issue that a `POST` request to `/{index}/_settings` is being treated as an index request when ideally it would match the update settings handler for `/{index}/_settings` and return a 405 since this endpoint only accepts `PUT` requests. Instead, it's matching the `POST` handler for `/{index}/{type}` and that's why the we see a mapper parsing exception here.

**Comment by AdityaJNair** [comment]
Hey do you mind if we can have a go at this issue?

**Comment by jasontedor** [comment]
> Hey do you mind if we can have a go at this issue?

There's an open PR to address this issue.

**Comment by olcbean** [comment]
I cannot reproduce this issue in `7.0` ( 95dea2408d4c6b70fe43ea3d4294dd98f89178b1 )
I think the current behavior ( shown below ) is correct and this issue can be closed.

```
curl -XPUT "http://localhost:9200/index1/_settings?pretty" -H 'Content-Type: application/json'
```
```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "parse_exception",
        "reason" : "request body is required"
      }
    ],
    "type" : "parse_exception",
    "reason" : "request body is required"
  },
  "status" : 400
}
```

```
curl -XPUT "http://localhost:9200/index1/_settings?pretty" -H 'Content-Type: application/json' -d '{}'
```
```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "action_request_validation_exception",
        "reason" : "Validation Failed: 1: no settings to update;"
      }
    ],
    "type" : "action_request_validation_exception",
    "reason" : "Validation Failed: 1: no settings to update;"
  },
  "status" : 400
}
```
Also using unsupported op `POST` is correctly detected and reported
```
curl -XPOST "http://localhost:9200/index1/_settings?pretty" -H 'Content-Type: application/json' -d '{}'
```
```
{
  "error" : "Incorrect HTTP method for uri [/index1/_settings?pretty] and method [POST], allowed: [PUT, GET]",
  "status" : 405
}
```

**Comment by javanna** [comment]
thanks @olcbean for testing this, I think it was fixed with #24437 but we forgot to close this issue. Closing.
